File: backend/src/main/java/com/springboot/backend/luismartinez/billingsapp/billingbackend/entities/Invoice.java
package com.springboot.backend.luismartinez.billingsapp.billingbackend.entities;

import com.springboot.backend.luismartinez.billingsapp.billingbackend.entities.enums.InvoiceStatus;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Data;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Setter
@Getter
@Entity
@Table(name = "invoices")
public class Invoice {

    // Getters and Setters
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "invoice_number", unique = true, nullable = false)
    private String invoiceNumber;

    @NotNull(message = "Creation date cannot be null")
    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "issue_date")
    private LocalDate issueDate;

    @Column(name = "due_date")
    private LocalDate dueDate;

    @Column(name = "paid_at")
    private LocalDateTime paidAt;

    private String currency = "EUR";

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    @OneToMany(mappedBy = "invoice", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<InvoiceItem> items = new ArrayList<>();;

    @Column(name = "subtotal", precision = 15, scale = 2)
    private BigDecimal subtotal;

    @Column(name = "tax_rate", precision = 5, scale = 4)
    private BigDecimal taxRate = BigDecimal.valueOf(0.19); // Default tax rate 19%

    @Column(name = "tax_amount", precision = 15, scale = 2)
    private BigDecimal taxAmount;

    @Column(name = "total", precision = 15, scale = 2)
    private BigDecimal total;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private InvoiceStatus status;

    @OneToMany(mappedBy = "invoice", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Payment> payments = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
    }

    public void addPayment(Payment payment) {
        payments.add(payment);
        payment.setInvoice(this);
    }

}

File: frontend/lminvoice/src/app/features/invoices/models/invoice.ts

export interface InvoiceItem {
  productId: number;
  productName?: string;
  quantity: number;
  price: number;
  total?: number;
}

export interface Invoice {
  id?: number;
  customerId?: number;
  items: InvoiceItem[];
  invoiceNumber?: string;
  createdAt?: string;
  issueDate?: string;
  dueDate?: string;
  paidAt?: string;
  subtotal?: number;
  taxAmount?: number;
  customer?: { id: number; name: string; email?: string };
  total?: number;
  status?: string;
}

export interface CreateInvoice {
  customer: { id: number };
  items: { product: { id: number }; quantity: number }[];
}

export type PaymentMethod = 'CASH' | 'TRANSFER' | 'CARD';

export interface Payment {
  id: number;
  amount: number;
  paidAt: string;
  method: PaymentMethod;
  invoiceId: number;
}

export interface CreatePayment {
  amount: number;
  method: PaymentMethod;
}
